src/DominantJuice.sol

```diff
@@ -0,0 +1,396 @@
+import {JBTokenAmount} from "@jbx-protocol/juice-contracts-v3/contracts/structs/JBTokenAmount.sol";
+import {Ownable} from "@openzeppelin/contracts/access/Ownable.sol";
+import {UD60x18, ud, add, pow, powu, div, mul, wrap, unwrap} from "@prb/math/UD60x18.sol";
+
+/**
+ * @title Dominant Juice
+ * @author Armand Daigle
+ * @notice This contract is a JB Data Source, JB Pay Delegate, JB Redemption Delegate, and a Refund Bonus Escrow
+ * implementation. Campaigns, referred to as cycles from here on, that use this implementation as a Data Source
+ * will incorporate Alex Tabarrok's dominant assurance mechanism, which incentivize early pledgers with a campaign
+ * creator-provided refund bonus that can be withdrawn from this contract after a failed campaign.
+ * @dev This contract uses the JB Data Source, Pay Delegate, and Pay Redemption interfaces and extends the
+ * functionality of a Juicebox project. There is no constructor since each campaign will deploy a clone of this
+ * contract. The initialize() function acts as the constructor. This contract is written to only be used for one cycle.
+ * If projects want dominant assurance on more cycles, refactoring is required.
+ */
+contract DominantJuice is IJBFundingCycleDataSource3_1_1, IJBPayDelegate3_1_1, IJBRedemptionDelegate3_1_1, Ownable {
+    /// Juicebox Contracts
+
+    IJBController3_1 public controller;
+    IJBDirectory public directory; // directory of terminals and controllers for projects.
+    IJBFundAccessConstraintsStore public fundAccessConstraintsStore;
+    IJBSingleTokenPaymentTerminalStore3_1_1 public paymentTerminalStore;
+    IJBSingleTokenPaymentTerminal public paymentTerminal;
+
+    /// Juicebox Project State Variables
+    uint256 public projectId;
+    uint256 public projectConfiguration;
+    address public paymentToken;
+
+    /// Dominant Assurance State Variables
+    uint256 public cycleExpiryDate;
+    uint256 public cycleTarget;
+    uint256 public minimumPledgeAmount;
+    uint256 public totalRefundBonus;
+    uint256 public totalAmountPledged;
+    uint256 public totalAmountPledged_fromJB;
+    UD60x18 public rateOfDecay = ud(0.75e18); // Need to confirm we want to hardcode this
+    bool public isCycleExpired = false;
+    bool public isTargetMet = false;
+    bool public hasCreatorWithdrawnAllFunds = false;
+
+    /// Pledger Mappings and Array
+    mapping(address => uint256) public pledgedAmount;
+    mapping(address => bool) public hasBeenRefunded;
+    mapping(uint256 => mapping(address => uint256)) public orderToPledgerToAmount;
+    address[] public pledgers;
+    uint32 public pledgeOrder; // make internal?
+
+    // Events for transparency to pledgers
+    event RefundBonusDeposited(address, uint256 indexed);
+    event CycleHasClosed(bool indexed, bool indexed);
+    event CycleRefundBonusWithdrawal(address indexed, uint256 indexed);
+    event OwnerWithdrawal(address, uint256);
+    event Test(string, uint256);
```

Comment:
what's this `Test` event for?

Reply:
Oh, sorry, I thought I had included a note on that in the review doc, but I did not. This is just a temporary test helper I just learned about to give you logic step feedback with variable info as well. Fairly handy actually. But yes, I will make sure it is deleted whenever I make pushes.